astr303a.py

```diff
@@ -0,0 +1,571 @@
+            sumden += 1/w
+        meanra = sumnum/sumden
+
+        #compute mean Declination
+        sumnum = 0
+        sumden = 0
+        for i in range(0,len(dataarray)):
+            w = 10**(dataarray[i][4]/2.5)
+            sumnum += dataarray[i][2]/w
+            sumden += 1/w
+        meandec = sumnum/sumden
+        
+        print('Found mean RA,Dec: ',meanra,meandec)
+        return meanra,meandec
+
+    #perform gaussian fitting of current galaxy set
+    def part8b(dataarray):
+        
+        #create list of all redshifts
+        redshift = []
+        for i in range(0,len(dataarray)):
+            redshift.append(dataarray[i][6])
+
+        #Gather data in bins of width 0.001. Retrive counts from each bin. 
+        histogram = py.hist(redshift,bins=np.arange(0.005, 0.05, 0.001))
+        counts = histogram[0]
+        shifts = histogram[1]
+        binset = []
+
+        #Create list of z values at bin centers
+        for i in range(0,len(shifts)-1):
+            binset.append(shifts[i]+0.0005)
+
+        #Get initial parameter guesses (mean,stddev). Get gaussian fit.
+        def gauss(x, a, x0, sigma):
+            return a*np.exp(-(x-x0)**2/(2*sigma**2))
+        stddev = np.std(redshift)
+        mean = np.mean(redshift)
+        coeffs, pcov = optimize.curve_fit(gauss, binset, counts, p0 = [1, mean, stddev])       
+        plt.clf()
+
+        print('Gaussian fit: peak,mean,sigma',coeffs)
+        return coeffs,mean,stddev
+
+    #Compute recessional velocity and distance of current galaxy set.
+    def part8c(coeffs,dataarray):
+        
+        #Compute recessional velocity 
+        zcoma = coeffs[1]
+        sigmacoma = coeffs[2]
+        c = 2.99792485e8
+        vcoma = c*(((zcoma+1)**2)-1)/(((zcoma+1)**2)+1)
+        vcoma = vcoma/1000
+
+        #Compute distance from Hubble law
+        H = 67.8
+        dcoma = vcoma/H
+        
+        print('Coma radial velocity: '+str(vcoma)+'km/s')
+        print('Coma distance (from Hubble law): '+str(dcoma)+'Mpc')
+        return dcoma,zcoma
+
+
+    dataarray2 = [] #create array that will contain galaxies removed from Coma Cluster member set.
+
+    ### SELECTION ALGOITHM ###
+    j=0 #interation number
+    cont=True
+    while cont==True: #repeat iterations until loop broken
+        dell=False #Set 'galaxy deleted' boolean to False. Convergence is reached when dell=False at the end of the interation.
+        numdell = 0 #Reset number of galaxies deleted on this iteration to zero
+        j+=1 #increase iteration number
+
+        print('\nIteration '+str(j)+' of member group selection...')
+        comaRA,comaDEC = part8a(dataarray) #calculate average RA,DEC
+        coeffs,mean,stddev = part8b(dataarray) #fit gaussian 
+        dcoma,zcoma = part8c(coeffs,dataarray) #get distance and average z
+        threesigma = abs(coeffs[2]*3) #set 3 sigma limit      
+
+        #check all galaxies against current values
+        i = 0 
+        while i<len(dataarray):
```

Comment:
There's a ton of opportunity here to make this code more declarative and make substantial speed ups through vectorizing your operations across the array (exactly what `numpy` is good at). Instead of using this while loop, and calling `projecteddistance()` in the middle of it, instead you could write projected distance to take two vectors of floats, and 3 constants and call it once (instead of 769 times). Instant speed up.